examples/ffa_simulation.rs

```diff
@@ -0,0 +1,224 @@
+ On every tick every entity/unit:
+    1. Moves in random direction.
+     2. Finds closest entity to itself.
+     3. Fires at it and applies damage.
+     4. Gets damaged by other entities firing at them.
+     5. If health <= 0, the unit dies.
+State of the simulation is displayed in the sconsole through println! functions.
+*/
+
+#[derive(Debug)]
+struct Position {
+    x: i32,
+    y: i32,
+}
+
+#[derive(Debug)]
+struct Health {
+    hp: i32,
+}
+
+#[derive(Debug)]
+struct Speed {
+    speed: i32,
+}
+
+#[derive(Debug)]
+struct Damage {
+    dmg: i32,
+}
+
+#[derive(Debug)]
+struct KillCount {
+    count: i32,
+}
+
+fn manhattan_dist(x0: i32, x1: i32, y0: i32, y1: i32) -> i32 {
+    let dx = (x0 - x1).max(x1 - x0);
+    let dy = (y0 - y1).max(y1 - y0);
+    return dx + dy;
+}
+
+// Spawns entities one by one. It's better to use spawn_batch instead, which is used in function below.
+fn _spawn_enitites(world: &mut World, n: usize) {
+    let mut rng = thread_rng();
+
+    for _ in 0..n {
+        let pos = Position {
+            x: rng.gen_range(-10, 10),
+            y: rng.gen_range(-10, 10),
+        };
+        let s = Speed {
+            speed: rng.gen_range(1, 5),
+        };
+        let hp = Health {
+            hp: rng.gen_range(30, 50),
+        };
+        let dmg = Damage {
+            dmg: rng.gen_range(1, 10),
+        };
+        let kc = KillCount { count: 0 };
+
+        world.spawn((pos, s, hp, dmg, kc));
+    }
+}
+
+fn batch_spawn_entities(world: &mut World, n: usize) {
+    let mut rng = thread_rng();
+
+    let to_spawn = (0..n).map(|_| {
+        let pos = Position {
+            x: rng.gen_range(-10, 10),
+            y: rng.gen_range(-10, 10),
+        };
+        let s = Speed {
+            speed: rng.gen_range(1, 5),
+        };
+        let hp = Health {
+            hp: rng.gen_range(30, 50),
+        };
+        let dmg = Damage {
+            dmg: rng.gen_range(1, 10),
+        };
+        let kc = KillCount { count: 0 };
+
+        (pos, s, hp, dmg, kc)
+    });
+
+    world.spawn_batch(to_spawn);
+}
+
+fn system_integrate_motion(world: &mut World) {
+    let mut rng = thread_rng();
+
+    for (id, (pos, s)) in &mut world.query::<(&mut Position, &Speed)>() {
+        let change = (
+            rng.gen_range(-s.speed, s.speed),
+            rng.gen_range(-s.speed, s.speed),
+        );
+        pos.x += change.0;
+        pos.y += change.1;
+        println!("Unit {:?} moved to {:?}", id, pos);
+    }
+}
+
+// In this system entities find the closest entity and fire at them
+fn system_fire_at_closest(world: &mut World) {
+    for (id0, (pos0, dmg0, kc0)) in
+        &mut world.query::<With<Health, (&Position, &Damage, &mut KillCount)>>()
+    {
+        let mut min_dist: Option<i32> = None;
+        let mut closest_id: Option<Entity> = None;
+
+        // find closest:
+        for (id1, pos1) in &mut world.query::<With<Health, &Position>>() {
```

Comment:
It's worth leaving a comment here that explains that nested queries are O(n^2) and you usually want to avoid that by using some sort of spatial index like a quadtree or more general BVH, which we don't bother with here since it's out of scope for the example.